apps/emqx_message_transformation/src/emqx_message_transformation_registry.erl

```diff
@@ -0,0 +1,245 @@
+    inc_failed/1,
+
+    start_link/0,
+    metrics_worker_spec/0
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2
+]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+-define(TRANSFORMATION_TOPIC_INDEX, emqx_message_transformation_index).
+-define(TRANSFORMATION_TAB, emqx_message_transformation_tab).
+
+-define(METRIC_NAME, message_transformation).
+-define(METRICS, [
+    'matched',
+    'succeeded',
+    'failed'
+]).
+-define(RATE_METRICS, ['matched']).
+
+-type transformation_name() :: binary().
+%% TODO
+-type transformation() :: #{atom() => term()}.
+-type position_index() :: pos_integer().
+
+-record(reindex_positions, {transformations :: [transformation()]}).
+-record(insert, {pos :: position_index(), transformation :: transformation()}).
+-record(update, {old :: transformation(), pos :: position_index(), new :: transformation()}).
+-record(delete, {transformation :: transformation()}).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec start_link() -> gen_server:start_ret().
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec lookup(transformation_name()) ->
+    {ok, transformation()} | {error, not_found}.
+lookup(Name) ->
+    case emqx_utils_ets:lookup_value(?TRANSFORMATION_TAB, Name, undefined) of
+        undefined ->
+            {error, not_found};
+        Transformation ->
+            {ok, Transformation}
+    end.
+
+-spec reindex_positions([transformation()]) -> ok.
+reindex_positions(Transformations) ->
+    gen_server:call(?MODULE, #reindex_positions{transformations = Transformations}, infinity).
+
+-spec insert(position_index(), transformation()) -> ok.
+insert(Pos, Transformation) ->
+    gen_server:call(?MODULE, #insert{pos = Pos, transformation = Transformation}, infinity).
+
+-spec update(transformation(), position_index(), transformation()) -> ok.
+update(Old, Pos, New) ->
+    gen_server:call(?MODULE, #update{old = Old, pos = Pos, new = New}, infinity).
+
+-spec delete(transformation()) -> ok.
+delete(Transformation) ->
+    gen_server:call(?MODULE, #delete{transformation = Transformation}, infinity).
+
+%% @doc Returns a list of matching transformation names, sorted by their configuration order.
+-spec matching_transformations(emqx_types:topic()) -> [transformation()].
+matching_transformations(Topic) ->
+    Transformations0 = [
+        {Pos, Transformation}
+     || M <- emqx_topic_index:matches(Topic, ?TRANSFORMATION_TOPIC_INDEX, [unique]),
+        [Pos] <- [emqx_topic_index:get_record(M, ?TRANSFORMATION_TOPIC_INDEX)],
+        {ok, Transformation} <- [
+            lookup(emqx_topic_index:get_id(M))
+        ]
+    ],
+    Transformations1 = lists:sort(
```

Comment:
maybe should index Pos and put Name as value so there is no need to sort.

Reply:
same for schema validation.

Reply:
:heavy_check_mark:  for message transformation, but I think refactoring this for schema validation should be done in a separate PR.

Reply:
https://github.com/emqx/emqx/pull/13221